src/Microsoft.Azure.ServiceBus/Core/MessageReceiver.cs

```diff
@@ -704,16 +707,36 @@ public void RegisterMessageHandler(Func<Message, CancellationToken, Task> handle
                 throw new ServiceBusException(true, Resources.AmqpFieldSessionId);
             }
 
+            receivingAmqpLink.Closed += this.OnSessionReceiverLinkClosed;
             this.SessionIdInternal = tempSessionId;
             long lockedUntilUtcTicks;
             this.LockedUntilUtcInternal = receivingAmqpLink.Settings.Properties.TryGetValue(AmqpClientConstants.LockedUntilUtc, out lockedUntilUtcTicks) ? new DateTime(lockedUntilUtcTicks, DateTimeKind.Utc) : DateTime.MinValue;
         }
 
+        void OnSessionReceiverLinkClosed(object sender, EventArgs e)
```

Comment:
>OnSessionReceiverLinkClosed [](start = 13, length = 27)

Instead cant we just call `this.CloseSession` and `closeAsync` here? Lets close the link completely, and then each operation should check `clientEntity.IsClosedOrClosing` to take actions .